coverage/lcovreport.py

```diff
@@ -0,0 +1,57 @@
+# Licensed under the Apache License: http://www.apache.org/licenses/LICENSE-2.0
+# For details: https://github.com/nedbat/coveragepy/blob/master/NOTICE.txt
+
+"""LCOV reporting for coverage.py"""
+
+import sys
+from hashlib import md5
+
+from coverage.report import get_analysis_to_report
+from coverage.exceptions import NoSource, NotPython
+
+
+class LcovReporter:
+    """A reporter for writing LCOV coverage reports"""
+
+    report_type = "LCOV report"
+
+    def __init__(self, coverage):
+        self.coverage = coverage
+        self.config = self.coverage.config
+
+    def report(self, morfs, outfile=None):
+        """Renders the full lcov report
+
+        'morfs' is a list of modules or filenames
+
+        outfile is the file object to write the file into
+        """
+
+        self.coverage.load()
+        self.coverage.get_data()
+        lcov_data = ""
+        outfile = outfile or sys.stdout
+        coverage_data = self.coverage.get_data()
+
+        for fr, analysis in get_analysis_to_report(self.coverage, morfs):
+            lcov_data += self.get_lcov(fr, analysis)
+        outfile.write(lcov_data)
+
+    def get_lcov(self, fr, analysis):
+        """Prooduces the lcov data for a single file"""
+        output = ""
+        token_lines = analysis.file_reporter.source().splitlines()
+        output += "TN:\n"
+        output += f"SF:{fr.relative_filename()}\n"
+
+        lines_hit = 0
+        for i, _ in enumerate(token_lines, 1):
+            if i in analysis.executed:
+                output += f"DA:{i},1\n"
+                lines_hit += 1
+            elif i in analysis.missing:
+                output += f"DA:{i},0\n"
```

Comment:
Good start! Now you may notice that `analysis.executed` and `analysis.missing` are lists of line numbers, and you're checking if line numbers are in those lists... By enumerating over the original source file, with all the file reading, memory and other overheads, without using any of the text in it.

Essentially, you're reading a file and then running enumerate on it to iterate over the number of lines in the file (note: *not* iterate over the lines in the file). When you already have all that data and don't even need to iterate.

How about something like this?

```suggestion
        for covered in analysis.executed:
            # Note: Coveragepy currently only supports checking *if* a line has been
            # executed, not how many times, so we set this to 1 for nice output even
            # if it's technically incorrect
            # TODO: Add checksum
            output += f"DA:{covered},1\n"
        for missed in analysis.missing:
            output += f"DA:{missed},0\n"
```

If it has to be ordered, this gets a bit more complex, but I highly doubt it considering the format supports specifying the line number.